packages/formula-editor/src/recommendor.ts

```diff
@@ -0,0 +1,127 @@
+export class Recommender {
+  private _trie: TrieNode;
+  private _minSuggestionLen: number;
+
+  constructor(variables: Map<string, number>, minSuggestionLen: number) {
+    this._minSuggestionLen = minSuggestionLen > 0 ? minSuggestionLen : 1;
+    this._trie = new TrieNode();
+    this._trie.insertAll();
+
+    for (let variable of variables) {
+      this.insert(variable[0]);
+    }
+  }
+
+  insert(
+    word: string,
+    position: number = -1,
+    node: TrieNode | undefined = undefined
+  ): void {
+    if (position == -1) {
+      this.insert(word, 0, this._trie);
+      return;
+    }
+
+    if (position == word.length) {
+      node?.addChild("\0");
+      return;
+    }
+
+    if (!node!.getChild(word[position])) {
+      node?.addChild(word[position]);
+    }
+
+    this.insert(word, position + 1, node!.getChild(word[position]));
```

Comment:
suggestion: can remove the use of _position_, by slicing out first character with each recursive call

Reply:
Not feasible as slice creates a new string, which will create memory overhead.

> The slice() method extracts a section of a string and returns it as a new string, without modifying the original string.
> [MDN Docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice)